oap-server/server-alarm-plugin/src/main/java/org/apache/skywalking/oap/server/core/alarm/provider/AlarmModuleProvider.java

```diff
@@ -49,19 +53,26 @@
         }
         RulesReader reader = new RulesReader(applicationReader);
         Rules rules = reader.readRules();
+        alarmRulesWatcher = new AlarmRulesWatcher(rules, this, this::reload);
+    }
+
+    private void reload(final Rules rules) {
```

Comment:
This `reload` is a hijack to existing code. `ModuleProvider#registerServiceImplementation` didn't expect runtime change in the design. So you could see `Map<Class<? extends Service>, Service> services` for service register, which is not thread-safe.

We need to reconsider this, even whether do we need to approve the service hot-switch